core/src/main/c/share/dedup.cpp

```diff
@@ -0,0 +1,644 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+#include "jni.h"
+#include <cstring>
+#include "util.h"
+#include "simd.h"
+#include "ooo_dispatch.h"
+#include "bit_vector.h"
+#include <algorithm>
+
+#pragma pack (push, 1)
+// Should match data structure described in DedupColumnCommitAddresses.java
+struct dedup_column {
+    int32_t column_type;
+    int32_t value_size_bytes;
+    int64_t column_top;
+    void *column_data;
+    void *o3_data;
+    int64_t reserved1;
+    int64_t reserved2;
+    int64_t reserved3;
+    char null_value[32];
+};
+#pragma pack(pop)
+
+template<typename T>
+struct dedup_column_t : dedup_column {
+    [[nodiscard]] inline const T *get_column_data() const {
+        return (T *) dedup_column::column_data;
+    };
+
+    [[nodiscard]] inline const T *get_o3_data() const {
+        return (T *) dedup_column::o3_data;
+    };
+
+    [[nodiscard]] inline const T &get_null_value() const {
+        return (T &) dedup_column::null_value;
+    };
+};
+
+struct int256 {
+    __int128 lo;
+    __int128 hi;
+};
+
+inline bool operator>(const int256 &a, const int256 &b) {
+    return a.hi > b.hi || a.lo > b.lo;
+}
+
+inline bool operator<(const int256 &a, const int256 &b) {
+    return a.hi < b.hi || a.lo < b.lo;
+}
+
+template<typename lambda_diff>
+inline int64_t branch_free_search(int64_t searchIndex, const index_t *array, const int64_t count, lambda_diff compare) {
+    const index_t *base = array;
+    int64_t n = count;
+    while (n > 1) {
+        int64_t half = n / 2;
+        MM_PREFETCH_T0(base + half / 2);
+        MM_PREFETCH_T0(base + half + half / 2);
+        auto diff = compare(searchIndex, base[half].i);
+        base = (diff > 0) ? base + half : base;
+        n -= half;
+    }
+    if (compare(searchIndex, base[0].i) == 0) {
+        return base - array;
+    }
+    if (base - array + 1 < count && compare(searchIndex, base[1].i) == 0) {
+        return base - array + 1;
+    }
+    return -1;
+}
+
+template<typename lambda_diff>
+int64_t merge_dedup_long_index_int_keys(
+        const uint64_t *src,
+        const int64_t src_lo,
+        const int64_t src_hi_incl,
+        index_t *const index,
+        const int64_t index_lo,
+        const int64_t index_hi_incl,
+        index_t *dest_index,
+        const lambda_diff compare
+) {
+    int64_t src_pos = src_lo;
+    int64_t index_pos = index_lo;
+    index_t *dest = dest_index;
+
+    bit_vector used_indexes = {};
+    while (src_pos <= src_hi_incl && index_pos <= index_hi_incl) {
+        // Perform normal merge until the timestamp matches.
+        if (src[src_pos] < index[index_pos].ts) {
+            dest[0].ts = src[src_pos];
+            dest[0].i = src_pos | (1ull << 63);
```

Comment:
src_pos is always positive, why not simply  dest[0].i = -src_pos; ?

Reply:
This is for existing row merge procs, like `merge_copy_var_column` in O3. These procs do 

```C
const uint32_t bit = (row >> 63);
const uint64_t rr = row & ~(1ull << 63);
```

and I don't want to change them for dedup puroses